### PR TITLE
[Edge] Elaborate edge API description.

### PR DESCRIPTION
**Pull request by gichan-jang**
- Add elaborate edge API description.
 - Add example code.

Signed-off-by: gichan <gichan2.jang@samsung.com>

### CODE REVIEW

File: include/nnstreamer-edge.h [hunk 1]
```diff
@@ -96,58 +96,199 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
 
 /**
  * @brief Start the nnstreamer edge.
+ * @param[in] edge_h The edge handle.
+ * @return 0 on success. Otherwise a negative error value.
+ * @retval #NNS_EDGE_ERROR_NONE Successful.
+ * @retval #NNS_EDGE_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #NNS_EDGE_ERROR_OUT_OF_MEMORY Failed to allocate required memory.
+ * @retval #NNS_EDGE_ERROR_INVALID_PARAMETER Given parameter is invalid.
+ * @retval #NNS_EDGE_ERROR_CONNECTION_FAILURE Failed to connect to destination.
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:159]: error when failed to set the available port. 
no connection in start()

File: include/nnstreamer-edge.h [hunk 2]
```diff
@@ -96,58 +96,199 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ * @retval #NNS_EDGE_ERROR_INVALID_PARAMETER Given parameter is invalid.
+ *
+ * Here is an example of the usage:
+ * @code
+ * static int _nns_edge_event_cb (nns_edge_event_h event_h, void *user_data) {
+ *   nns_edge_event_e event_type;
+ *   nns_edge_event_get_type (event_h, &event_type);
+ *   switch (event_type) {
+ *     case NNS_EDGE_EVENT_CAPABILITY:
+ *     case NNS_EDGE_EVENT_NEW_DATA_RECEIVED:
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:116]: recommend to add functions - nns_edge_event_parse_new_data() and nns_edge_event_parse_capability()
```
case NNS_EDGE_EVENT_CAPABILITY:
{
  char *capability;
  nns_edge_event_parse_capability (event_h, &capability);
  /** 
   * Check capability string and return NNS_EDGE_ERROR_NONE if capability is acceptable.
   * Returning error code will deny the connection.
   */
}
case NNS_EDGE_EVENT_NEW_DATA_RECEIVED:
{
  nns_edge_data_h data_h;
  nns_edge_event_parse_new_data (event_h, &data_h);
  /**
   * You should release returned data_h using nns_edge_data_destroy().
   * Recommend not to spend too much time in event callback.
   */
}
```

File: include/nnstreamer-edge.h (hunk 1 above)
Comment (gichan-jang) [reply include/nnstreamer-edge.h]: Thanks. Updated.

File: include/nnstreamer-edge.h (hunk 2 above)
Comment (gichan-jang) [reply include/nnstreamer-edge.h:128]: Thanks. Updated.

File: include/nnstreamer-edge.h [hunk 3]
```diff
@@ -96,119 +96,360 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ *
+ *       // You should release returned data_h using nns_edge_data_destroy().
+ *       // Recommend not to spend too much time in event callback.
+ *       nns_edge_data_destroy (data_h);
+ *       break;
+ *     }
+ *     default:
+ *       break;
+ *   }
+ *   return 0;
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:141]: plz update return code to NNS_EDGE_ERROR_NONE

File: include/nnstreamer-edge.h [hunk 4]
```diff
@@ -96,119 +96,360 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
 
 /**
  * @brief Create edge handle.
+ * @param[in] id Unique id in local network
+ * @param[in] connect_type value of nns_edge_flag_e. Connection type between edge nodes.
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:100]: nns_edge_flag_e > nns_edge_connect_type_e

File: include/nnstreamer-edge.h [hunk 5]
```diff
@@ -96,119 +96,360 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ * // [Optional] Set edge info if necessary.
+ * ret = nns_edge_set_info (edge_h, "HOST", "127.0.0.1");
+ *
+ * // Set event callback to receive the new data or capability.
+ * // Query-server, query-client and subscriber need to set event callback.
+ * nns_edge_set_event_callback (src->edge_h, _nns_edge_event_cb, src);
+ *
+ * // start edge handle.
+ * ret = nns_edge_start (edge_h);
+ *
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:162]: // Connect to destination node.
ret = nns_edge_connect (edge_h, DEST_IP, DEST_PORT);

Comment (jaeyun-jung) [reply include/nnstreamer-edge.h:162]: uh...it is described below, ignore this comment.

File: include/nnstreamer-edge.h (hunk 4 above)
Comment (gichan-jang) [reply include/nnstreamer-edge.h]: Updated. Thanks.

File: include/nnstreamer-edge.h [hunk 6]
```diff
@@ -96,119 +96,360 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ * @return 0 on success. Otherwise a negative error value.
+ * @retval #NNS_EDGE_ERROR_NONE Successful.
+ * @retval #NNS_EDGE_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #NNS_EDGE_ERROR_INVALID_PARAMETER Given parameter is invalid
  */
 int nns_edge_event_parse_capability (nns_edge_event_h event_h, char **capability);
 
 /**
  * @brief Create nnstreamer edge data.
+ * @param[out] data_h Handle of edge data. data handle should be released using nns_edge_data_destroy().
```
Comment (jaeyun-jung) [inline include/nnstreamer-edge.h:359]: line 359 and 381
`Caller should release returned handle using nns_edge_data_destroy().`

**Review by jaeyun-jung — APPROVED** [review]
LGTM. Please update release function (param[out] edge-data).

File: include/nnstreamer-edge.h [hunk 7]
```diff
@@ -96,119 +96,362 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ * @retval #NNS_EDGE_ERROR_NONE Successful.
+ * @retval #NNS_EDGE_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #NNS_EDGE_ERROR_INVALID_PARAMETER Given parameter is invalid.
+ * @section table_sec Edge info table
+ * key                  | value
+ * ---------------------|--------------------------------------------------------------
+ * CAPS or CAPABILITY   | capapility strings.
+ * IP or HOST           | IP address of the node to accept connection from other node.
+ * PORT                 | Port of the node to accept connection from other node.
+ * DEST_IP or DEST_HOST | IP address of the destination node.
```
Comment (myungjoo) [inline include/nnstreamer-edge.h:302]: describe the usage for the case we have broker (MQTT) or broker-discovery (AITT).

File: include/nnstreamer-edge.h [hunk 8]
```diff
@@ -96,119 +96,362 @@ typedef void (*nns_edge_data_destroy_cb) (void *data);
...
+ * ret = nns_edge_create_handle ("TEMP_ID", NNS_EDGE_CONNECT_TYPE_TCP, NNS_EDGE_FLAG_RECV | NNS_EDGE_FLAG_SEND | NNS_EDGE_FLAG_SERVER, &edge_h);
+ * if (NNS_EDGE_ERROR_NONE != ret)
+ *   // Handle error case. After this, handling error case is repeated, so it is omitted.
+ *
+ * // [Optional] Set edge info if necessary.
+ * ret = nns_edge_set_info (edge_h, "HOST", "127.0.0.1");
+ *
+ * // Set event callback to receive the new data or capability.
+ * // Query-server, query-client and subscriber need to set event callback.
+ * nns_edge_set_event_callback (src->edge_h, _nns_edge_event_cb, src);
```
Comment (myungjoo) [inline include/nnstreamer-edge.h:158]: We need cases for:
1. query-server (sink-side & src-side)
2. query-client
3. stream-sink (or pub)
4. stream-src (or sub)

Note that the usage of the term "pub" in query-client is very confusing when we have 3 and 4.

**Comment by myungjoo** [comment]
Need more description on the function "brief" to let users understand which function to use.
E.g., "Create edge handle" --> "Create a handle representing an instance of edge-AI connection between a server and client (query) or a data publisher and scriber" The description should be descriptive enough for people who have not been working for nnstreamer.

**Comment by myungjoo** [comment]
Merge this after
1. update things that do not incur changes in other modules.
2. write changes (affecting other modules) that are to be done later (after M2 release) as an backlog issue at nnstreamer.git.

**Review by myungjoo — APPROVED** [review]

**Pull request merged**